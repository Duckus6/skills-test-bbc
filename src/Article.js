import React from 'react';
import * as PropTypes from 'prop-types';
import ArticleElements from './ArticleElements';
import styles from "./Article.module.css"

const Article = ({article}) => {

    const formatArticleElement = (articleElement)=>{
        const{type,model}=articleElement;
        const Component =ArticleElements[type];
        if (!Component){
            return (<p className={'error'}>An error has occurred</p>)
        }
        return <Component className={styles.component} {...model}/>
    };

    return(<div>
        {!!article && <div> {article.body.map(articleElement=>formatArticleElement(articleElement))}</div>}
    </div>)
};




Article.propTypes = {
    article: PropTypes.object.isRequired
};
export default Article;