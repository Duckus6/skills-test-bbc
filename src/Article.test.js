import React from 'react';
import axios from 'axios';
import AxiosMockAdapter from 'axios-mock-adapter';
import {mockArticle,mockArticles,mockInvalidArticle} from "./fixtures/articles";
import Article from "./Article";
import {mount} from "enzyme/";
import {act} from "@testing-library/react/";



const mock = new AxiosMockAdapter(axios);


describe("the server returns 200 code",()=>{
    beforeAll(()=>{
        mock.onGet(`data/${mockArticles[0]}`).reply(200,
            mockArticle
        );
        mock.onGet(`data/${mockArticles[1]}`).reply(200,
            mockInvalidArticle
        );
    });
    afterAll(()=>{
        mock.reset();
    });
    test('renders', async (done) => {
        jest.useFakeTimers();
        const wrapper = mount(<Article articleID={mockArticles[0]}/>);
        await act(async ()=>{
            jest.runAllImmediates();
        });
        wrapper.update();
        console.log(wrapper.debug());
        expect(wrapper.debug()).toMatchSnapshot();
        done()
    });
    test('renders error when invalid element is received', async (done) => {
        jest.useFakeTimers();
        const wrapper = mount(<Article articleID={mockArticles[1]}/>);
        await act(async ()=>{
            jest.runAllImmediates();
        });
        wrapper.update();
        console.log(wrapper.debug());
        expect(wrapper.containsMatchingElement(<p>An error has occurred</p>)).toBeTruthy();
        done();
    })
});


describe("the server returns 500 error",()=>{
    beforeAll(()=>{
        const error = '[Error: Request failed with status code 500]';
        mock.onGet(`data/${mockArticles[0]}`).reply(500,
            error
        );
    });
    afterAll(()=>{
        mock.reset();
    });
    test('renders', () => {
        const wrapper = mount(<Article articleID={mockArticles[0]}/>);
        expect(wrapper.debug()).toMatchSnapshot();
    })
});



