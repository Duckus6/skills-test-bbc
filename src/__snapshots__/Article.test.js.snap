// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the server returns 200 code renders 1`] = `
"<Article articleID=\\"article1.json\\">
  <div>
    <div>
       
      <Heading text=\\"Lorem ipsum dolor sit amet, consectetur adipiscing elit\\">
        <h4>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit
        </h4>
      </Heading>
      <Paragraph text=\\"Etiam ornare nulla in sem pharetra, vel varius magna tempus. Praesent pulvinar eget dolor vitae elementum. Donec id velit at sem gravida tristique non vitae lorem.\\">
        <p>
          Etiam ornare nulla in sem pharetra, vel varius magna tempus. Praesent pulvinar eget dolor vitae elementum. Donec id velit at sem gravida tristique non vitae lorem.
        </p>
      </Paragraph>
      <Image url=\\"https://avatars1.githubusercontent.com/u/16746643?s=460&v=4\\" altText=\\"Vestibulum pellentesque laoreet urna, eget dignissim lorem maximus vel\\" height=\\"420\\" width=\\"640\\">
        <img src=\\"https://avatars1.githubusercontent.com/u/16746643?s=460&v=4\\" alt=\\"Vestibulum pellentesque laoreet urna, eget dignissim lorem maximus vel\\" height=\\"420\\" width=\\"640\\" />
      </Image>
      <List type=\\"unordered\\" items={{...}}>
        <ul>
          <li>
            Vestibulum viverra viverra ullamcorper
          </li>
          <li>
            Aenean ut felis hendrerit, scelerisque est sed, rhoncus sem
          </li>
        </ul>
      </List>
      <List type=\\"ordered\\" items={{...}}>
        <ol>
          <li>
            Vestibulum viverra viverra ullamcorper
          </li>
          <li>
            Aenean ut felis hendrerit, scelerisque est sed, rhoncus sem
          </li>
        </ol>
      </List>
    </div>
  </div>
</Article>"
`;

exports[`the server returns 500 error renders 1`] = `
"<Article articleID=\\"article1.json\\">
  <div />
</Article>"
`;
